File: server/src/services/cache.ts
import type { Redis } from "ioredis";
import { DatabaseService } from "./database";

type CacheValue = string;

// Redis 可用时优先使用；Redis 不可用时自动回退到 MariaDB（user_sessions 表）
export class CacheService {
  private readonly db: DatabaseService;
  private readonly redis: Redis | null;
  private redisHealthy: boolean;

  constructor(deps: { db: DatabaseService; redis?: Redis | null }) {
    this.db = deps.db;
    this.redis = deps.redis ?? null;
    this.redisHealthy = !!this.redis;
  }

  async get(key: string): Promise<CacheValue | null> {
    if (this.redis && this.redisHealthy) {
      try {
        const val = await this.redis.get(key);
        if (val !== null) return val;
      } catch (error) {
        this.markRedisUnhealthy(error);
      }
    }

    if (key.startsWith("session_")) {
      return await this.getSessionFromDB(key);
    } else if (key.startsWith("sub_token_")) {
      return await this.getSubscriptionTokenFromDB(key);
    }

    return null;
  }

  async set(key: string, value: CacheValue, ttlSeconds = 300): Promise<void> {
    if (this.redis && this.redisHealthy) {
      try {
        await this.redis.set(key, value, "EX", ttlSeconds);
      } catch (error) {
        this.markRedisUnhealthy(error);
      }
    }

    if (key.startsWith("session_")) {
      await this.setSessionToDB(key, value, ttlSeconds);
    }
  }

  async delete(key: string): Promise<void> {
    if (this.redis && this.redisHealthy) {
      try {
        await this.redis.del(key);
      } catch (error) {
        this.markRedisUnhealthy(error);
      }
    }

    if (key.startsWith("session_")) {
      const token = key.replace("session_", "");
      await this.db.db
        .prepare("DELETE FROM user_sessions WHERE token = ?")
        .bind(token)
        .run();
    }
  }

  async cleanupExpiredSessions(): Promise<void> {
    await this.db.db
      .prepare('DELETE FROM user_sessions WHERE expires_at < CURRENT_TIMESTAMP')
      .run();
  }

  async deleteByPrefix(prefix: string): Promise<void> {
    if (this.redis && this.redisHealthy) {
      try {
        let cursor = "0";
        do {
          const [next, keys] = await this.redis.scan(cursor, "MATCH", `${prefix}*`, "COUNT", 100);
          cursor = next;
          if (keys.length) {
            await this.redis.del(...keys);
          }
        } while (cursor !== "0");
      } catch (error) {
        this.markRedisUnhealthy(error);
      }
    }

    if (prefix.startsWith("user_") && prefix.includes("_")) {
      const userId = Number.parseInt(prefix.split("_")[1] ?? "", 10);
      if (!Number.isNaN(userId)) {
        await this.db.db
          .prepare("DELETE FROM user_sessions WHERE user_id = ?")
          .bind(userId)
          .run();
      }
    }
  }

  private async getSessionFromDB(key: string): Promise<string | null> {
    const token = key.replace("session_", "");
    const stmt = this.db.db.prepare(`
      SELECT user_data FROM user_sessions 
      WHERE token = ? AND expires_at > CURRENT_TIMESTAMP
    `);
    const result = await stmt.bind(token).first<{ user_data: string }>();
    return result?.user_data ?? null;
  }

  private async setSessionToDB(key: string, value: string, ttlSeconds: number): Promise<void> {
    const token = key.replace("session_", "");
    const userData = JSON.parse(value);
    const expiresAt = new Date(Date.now() + ttlSeconds * 1000);

    await this.db.db
      .prepare(
        `
        INSERT INTO user_sessions (token, user_id, user_data, expires_at, created_at)
        VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP)
        ON DUPLICATE KEY UPDATE 
          user_id = VALUES(user_id),
          user_data = VALUES(user_data),
          expires_at = VALUES(expires_at)
      `
      )
      .bind(token, userData.id, value, expiresAt.toISOString().slice(0, 19).replace("T", " "))
      .run();
  }

  private async getSubscriptionTokenFromDB(key: string): Promise<string | null> {
    const token = key.replace("sub_token_", "");
    const stmt = this.db.db.prepare(`
      SELECT id AS user_id FROM users WHERE token = ?
    `);
    const result = await stmt.bind(token).first<{ user_id: number | string }>();
    if (result) {
      const userId = result.user_id.toString();
      if (this.redis && this.redisHealthy) {
        try {
          await this.redis.set(key, userId, "EX", 300);
        } catch (error) {
          this.markRedisUnhealthy(error);
        }
      }
      return userId;
    }
    return null;
  }

  private markRedisUnhealthy(error: unknown) {
    const message = error instanceof Error ? error.message : String(error);
    console.warn("[redis] fallback to MariaDB:", message);
    this.redisHealthy = false;
  }
}

File: server-rs/src/cache.rs
use chrono::{Duration, Utc};
use redis::AsyncCommands;
use serde_json::Value;
use sqlx::Row;

use crate::state::AppState;

fn redis_key(state: &AppState, key: &str) -> String {
  let prefix = state.env.redis_prefix.as_str();
  if prefix.is_empty() || key.starts_with(prefix) {
    key.to_string()
  } else {
    format!("{prefix}{key}")
  }
}

#[allow(dead_code)]
pub async fn cache_get(state: &AppState, key: &str) -> Option<String> {
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let redis_key = redis_key(state, key);
    let result: redis::RedisResult<Option<String>> = conn.get(redis_key).await;
    if let Ok(Some(value)) = result {
      return Some(value);
    }
  }

  if key.starts_with("session_") {
    return get_session_from_db(state, key).await;
  }

  if key.starts_with("sub_token_") {
    return get_subscription_from_db(state, key).await;
  }

  None
}

pub async fn cache_set(state: &AppState, key: &str, value: &str, ttl_seconds: u64) {
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let redis_key = redis_key(state, key);
    let _ = conn.set_ex::<_, _, ()>(redis_key, value, ttl_seconds).await;
  }

  if key.starts_with("session_") {
    let _ = set_session_to_db(state, key, value, ttl_seconds).await;
  }
}

pub async fn cache_delete(state: &AppState, key: &str) {
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let redis_key = redis_key(state, key);
    let _ = conn.del::<_, i32>(redis_key).await;
  }

  if key.starts_with("session_") {
    let token = key.trim_start_matches("session_");
    let _ = sqlx::query("DELETE FROM user_sessions WHERE token = ?")
      .bind(token)
      .execute(&state.db)
      .await;
  }
}

pub async fn cache_set_redis_only(state: &AppState, key: &str, value: &str, ttl_seconds: u64) {
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let redis_key = redis_key(state, key);
    let _ = conn.set_ex::<_, _, ()>(redis_key, value, ttl_seconds).await;
  }
}

pub async fn cache_get_redis_only(state: &AppState, key: &str) -> Option<String> {
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let redis_key = redis_key(state, key);
    let result: redis::RedisResult<Option<String>> = conn.get(redis_key).await;
    if let Ok(value) = result {
      return value;
    }
  }
  None
}

pub async fn cache_delete_by_prefix(state: &AppState, prefix: &str) {
  let redis = match state.redis.clone() {
    Some(conn) => conn,
    None => return
  };
  let mut conn = redis;
  let base = redis_key(state, prefix);
  let pattern = format!("{base}*");
  let mut cursor: u64 = 0;

  loop {
    let scan_result: redis::RedisResult<(u64, Vec<String>)> = redis::cmd("SCAN")
      .arg(cursor)
      .arg("MATCH")
      .arg(&pattern)
      .arg("COUNT")
      .arg(200)
      .query_async(&mut conn)
      .await;
    let (next_cursor, keys) = match scan_result {
      Ok(value) => value,
      Err(_) => break
    };
    if !keys.is_empty() {
      let _: redis::RedisResult<()> = redis::cmd("DEL").arg(keys).query_async(&mut conn).await;
    }
    if next_cursor == 0 {
      break;
    }
    cursor = next_cursor;
  }
}

#[allow(dead_code)]
async fn get_session_from_db(state: &AppState, key: &str) -> Option<String> {
  let token = key.trim_start_matches("session_");
  let row = sqlx::query(
    r#"
    SELECT user_data FROM user_sessions
    WHERE token = ? AND expires_at > CURRENT_TIMESTAMP
    "#
  )
  .bind(token)
  .fetch_optional(&state.db)
  .await
  .ok()?;
  row.and_then(|r| r.try_get::<Option<String>, _>("user_data").ok().flatten())
}

async fn set_session_to_db(
  state: &AppState,
  key: &str,
  value: &str,
  ttl_seconds: u64
) -> Result<(), String> {
  let token = key.trim_start_matches("session_");
  let user_data: Value = serde_json::from_str(value).map_err(|err| err.to_string())?;
  let user_id = user_data
    .get("id")
    .and_then(Value::as_i64)
    .unwrap_or(0);
  let expires_at = (Utc::now() + Duration::seconds(ttl_seconds as i64)).format("%F %T").to_string();

  sqlx::query(
    r#"
    INSERT INTO user_sessions (token, user_id, user_data, expires_at, created_at)
    VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP)
    ON DUPLICATE KEY UPDATE
      user_id = VALUES(user_id),
      user_data = VALUES(user_data),
      expires_at = VALUES(expires_at)
    "#
  )
  .bind(token)
  .bind(user_id)
  .bind(value)
  .bind(expires_at)
  .execute(&state.db)
  .await
  .map_err(|err| err.to_string())?;

  Ok(())
}

#[allow(dead_code)]
async fn get_subscription_from_db(state: &AppState, key: &str) -> Option<String> {
  let token = key.trim_start_matches("sub_token_");
  let row = sqlx::query("SELECT id AS user_id FROM users WHERE token = ?")
    .bind(token)
    .fetch_optional(&state.db)
    .await
    .ok()?;
  let user_id = row.and_then(|r| r.try_get::<Option<i64>, _>("user_id").ok().flatten())?;
  if let Some(redis) = state.redis.clone() {
    let mut conn = redis;
    let _ = conn
      .set_ex::<_, _, ()>(redis_key(state, key), user_id.to_string(), 300u64)
      .await;
  }
  Some(user_id.to_string())
}
